tests/end-to-end/upgrade/pkg/tests/function/function.go

```diff
@@ -0,0 +1,277 @@
+package function
+
+import (
+	"bytes"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"strings"
+	"time"
+
+	"github.com/google/uuid"
+	kubelessV1 "github.com/kubeless/kubeless/pkg/apis/kubeless/v1beta1"
+	kubeless "github.com/kubeless/kubeless/pkg/client/clientset/versioned"
+	kymaApi "github.com/kyma-project/kyma/components/api-controller/pkg/apis/gateway.kyma-project.io/v1alpha2"
+	kyma "github.com/kyma-project/kyma/components/api-controller/pkg/clients/gateway.kyma-project.io/clientset/versioned"
+	"github.com/sirupsen/logrus"
+	corev1 "k8s.io/api/core/v1"
+	extensionsv1 "k8s.io/api/extensions/v1beta1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	instr "k8s.io/apimachinery/pkg/util/intstr"
+	"k8s.io/client-go/kubernetes"
+	restclient "k8s.io/client-go/rest"
+)
+
+// LambdaFunctionUpgradeTest tests the creation of a kubeless function and execute a http request to the exposed api of the function after Kyma upgrade phase
+type LambdaFunctionUpgradeTest struct {
+	functionName, uuid string
+	kubelessClient     *kubeless.Clientset
+	coreClient         *kubernetes.Clientset
+	apiClient          *kyma.Clientset
+	nSpace             string
+	hostName           string
+}
+
+func int32Ptr(i int32) *int32 { return &i }
+
+// NewLambdaFunctionUpgradeTest returns new instance of the FunctionUpgradeTest
+func NewLambdaFunctionUpgradeTest(config *restclient.Config) *LambdaFunctionUpgradeTest {
+
+	kubelessClient, err := kubeless.NewForConfig(config)
+	if err != nil {
+		return &LambdaFunctionUpgradeTest{}
+	}
+
+	coreClient, err := kubernetes.NewForConfig(config)
+	if err != nil {
+		return &LambdaFunctionUpgradeTest{}
+	}
+
+	apiClient, err := kyma.NewForConfig(config)
+	if err != nil {
+		return &LambdaFunctionUpgradeTest{}
+	}
+
+	domainName := os.Getenv("DOMAIN")
+	if len(domainName) == 0 {
+		domainName = "kyma.local"
+	}
+
+	nSpace := strings.ToLower("LambdaFunctionUpgradeTest")
+	functionName := "hello"
+	hostName := fmt.Sprintf("%s-%s.%s", functionName, nSpace, domainName)
+	return &LambdaFunctionUpgradeTest{
+		kubelessClient: kubelessClient,
+		coreClient:     coreClient,
+		functionName:   functionName,
+		uuid:           uuid.New().String(),
+		nSpace:         nSpace,
+		hostName:       hostName,
+		apiClient:      apiClient,
+	}
+}
+
+// CreateResources creates resources needed for e2e upgrade test
+func (f *LambdaFunctionUpgradeTest) CreateResources(stop <-chan struct{}, log logrus.FieldLogger, namespace string) error {
```

Comment:
due to the way the upgrade tests are executed this `CreateResources` is the only function that will be executed before the upgrade attempt. imho we should check that these newly created resources are valid (we could just call the TestResources function here as well)

Reply:
As I understand the upgrade happens and then the resources are created and tested.
The way how the backup test works is better because we expect the same resource before and after.

Reply:
You can add here explicit checks for resources which can be not in a ready state and just return an error if they weren't. 
Resources are created before upgrade and tested after - it's described in the readme `usage` section. You can also run this test without doing an upgrade to check if your test case configuration is correct.